### PR TITLE
Add command for starting an application

### PR DESCRIPTION
**Pull request by matthiasbeyer**
This is WIP and targets #425.

@neithernut please review my exec() utility function, which is the first
step towards a command.

### CODE REVIEW

File: src/util/exec.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+#include "util/attributes.h"
+
+int
+ws_exec(
+    const char* call,
+    char* const argv[]
+)
+__ws_nonnull__(1)
+__ws_noreturn__
```
Comment (neithernut) [inline src/util/exec.h]: The function _will_ return!

**Comment by neithernut** [comment]
Please add a function to the global context in `src/action/context.c`.

File: src/util/exec.h (hunk 1 above)
Comment (matthiasbeyer) [inline src/util/exec.h]: Not always. If the `exec()` call works, it does not return (as stated in the manpage of the call).

**Comment by matthiasbeyer** [comment]
@neithernut Added, please review. I don't know whether this implementation is sufficient.

What I'm doing: Taking N arguments from the stack, which _all_ have to be strings. If there is a non-string argument, I `-EINVAL`. I collect the actual `char*`s from that strings and pass them to the new `ws_exec()` function as arguments.

No `NULL` checking by now.

File: src/util/exec.h (hunk 1 above)
Comment (neithernut) [inline src/util/exec.h]: The attribute is intended for functions that **never** return!

Comment (matthiasbeyer) [inline src/util/exec.h]: TIL.

File: src/action/context.c [hunk 2]
```diff
@@ -116,3 +127,41 @@ func_log(
...
+
+    while (n) {
+        struct ws_string* arg = ws_value_string_get(&stack[n + 1].string);
+        args[n] = ws_string_raw(arg);
+
+        --n;
+    }
+
+    return ws_exec(cmd, (char**)args);
+}
```
Comment (neithernut) [inline src/action/context.c]: The elements in `args` must be freed.
Also: add a blank line at EOF.

Comment (matthiasbeyer) [inline src/action/context.c]: Seriousely? The elements from `args` come directly out of a `struct ws_string`, which does the `free()` call when it gets cleaned up!

Comment (neithernut) [inline src/action/context.c]: `ws_string_raw` allocates memory for the string it returns. You have to free it!

**Comment by neithernut** [comment]
The first two arguments you get are the object and the command name. You have to ignore them.

File: src/action/context.c [hunk 3]
```diff
@@ -116,3 +127,43 @@ func_log(
...
+    enum ws_value_type iter_type;
+
+    // count number of arguments and do type checking meanwhile
+    for (n = 2, iter_type = ws_value_get_type(&stack[n + 1].value); iter; n++) {
+        if (iter_type != WS_VALUE_TYPE_NIL &&
+                iter_type != WS_VALUE_TYPE_STRING) {
+            return -EINVAL;
+        }
+
+        iter = iter_type != WS_VALUE_TYPE_NIL;
```
Comment (neithernut) [inline src/action/context.c]: You _really_ want to look for `WS_VALUE_TYPE_NONE`, and **not** for `WS_VALUE_TYPE_NIL`!

File: src/action/context.c [hunk 4]
```diff
@@ -116,3 +127,43 @@ func_log(
...
+    }
+
+    const char* args[n];
+
+    while (n) {
+        struct ws_string* arg = ws_value_string_get(&stack[n + 1].string);
+        args[n] = ws_string_raw(arg);
+
+        --n;
+    }
```
Comment (neithernut) [inline src/action/context.c]: As n is _offset_ by `2`, this will _definitely_ not result in what you want.

File: src/action/context.c (hunk 3 above)
Comment (matthiasbeyer) [inline src/action/context.c]: Indeed.

File: src/action/context.c (hunk 4 above)
Comment (matthiasbeyer) [inline src/action/context.c]: Indeed. Was fixed.

Comment (neithernut) [inline src/action/context.c]: No, you did **not** fix it, since you still pass `args` to `ws_exec` **without any offset!**

Comment (matthiasbeyer) [inline src/action/context.c]: `args` does not _have_ any offset, as it is a simple `char*[]`. There is no _need_ for an offset here. I compile the `ws_string`s from the arguments into a `char*[]`. In the arguments of the command, I have to use the offset, but not in the `char*[]`.

Comment (neithernut) [inline src/action/context.c]: Yep, now **please** read the thing you wrote again.

File: src/action/context.c [hunk 5]
```diff
@@ -116,3 +127,43 @@ func_log(
...
+
+    const char* args[n];
+
+    while (n > 2) {
+        struct ws_string* arg = ws_value_string_get(&stack[n + 1].string);
+        args[n - 2] = ws_string_raw(arg);
+
+        --n;
+    }
```
Comment (neithernut) [inline src/action/context.c]: Your code is still faulty. You need to decrement `n` _before_ copying.

**Comment by matthiasbeyer** [comment]
I did a complete reimplementation of the `func_exec()` function, as the code became a mess. Please review again.

File: src/action/context.c [hunk 6]
```diff
@@ -116,3 +127,42 @@ func_log(
...
+
+    const char* args[n];
+
+    while (n) {
+        --n;
+        struct ws_string* arg = ws_value_string_get(&stack[n].string);
+        args[n] = ws_string_raw(arg);
+    }
+
+    return ws_exec(cmd, (char**)args);
```
Comment (neithernut) [inline src/action/context.c]: The elements of `args` are still not `free()`d.

Comment (matthiasbeyer) [inline src/action/context.c]: The elements of `args` are still directly retreived from a `ws_string` which does the cleanup itself.

Comment (neithernut) [inline src/action/context.c]: **NO!!!** They are allocated in `ws_string_raw`, for god's sake!

File: src/action/context.c [hunk 7]
```diff
@@ -116,3 +127,51 @@ func_log(
...
+    stack += 2; // We don't care about object itself and command name
+
+    if (ws_value_get_type(&stack->value) != WS_VALUE_TYPE_STRING) {
+        return -EINVAL;
+    }
+
+    struct ws_string* strcmd    = ws_value_string_get(&stack->string);
+    const char* cmd             = ws_string_raw(strcmd);
+
+    stack++; // because of `cmd` is extracted now
```
Comment (xVin) [inline src/action/context.c]: "of" can be removed

File: src/action/context.c [hunk 8]
```diff
@@ -116,3 +127,51 @@ func_log(
...
+    enum ws_value_type iter_type;
+
+    // count number of arguments and do type checking meanwhile
+    for (n = 0, iter_type = ws_value_get_type(&stack[n].value); iter; n++) {
+        if (iter_type != WS_VALUE_TYPE_STRING) {
+            return -EINVAL;
+        }
+
+        iter = iter_type != WS_VALUE_TYPE_NONE;
+    }
```
Comment (neithernut) [inline src/action/context.c]: `iter_type` will _not_ be refreshed on every loop iteration (which is definitely not what you want).

**Comment by neithernut** [comment]
Is this still WIP?

**Comment by matthiasbeyer** [comment]
Nope

**Comment by neithernut** [comment]
Then ACK for squashing.

**Pull request merged**